### ISSUE TITLE
Some Fundamentals about building npm packages

### ISSUE THREAD

**Issue by reboottime** [reboottime/my-turbo-packages#16]
## Good practices on SDK generation

[source](https://nordicapis.com/10-best-practices-for-sdk-generation/)

 - User Centric: consider the developer
    - consider the  usage scenarios
    - including logging, let your developers know how your API is performing
        - requirements:
            - any error messages, 
            - debug messages, 
            - or notable events that occur during the execution of a code. 
        - example: [MSAL.js](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-common/src)
 - Standardization
    - naming convention
    - Transform Endpoints Into SDK Methods

**Comment by reboottime** [comment]
## Essential for Publishing NPM Packages


###  package.json essentials

- files: [The optional files field is an array of file patterns that describes the entries to be included when your package is installed as a dependency. ](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#files)

<br />

- exports: declares which module should be used when using module requests like import "package" or import "package/sub/path" , take bellow configuration as an example (original [source](https://github.com/emilkowalski/sonner/blob/main/package.json#L5)):
   ```json
       "exports": {
        "types": "./dist/index.d.ts",
        "import": "./dist/index.mjs",
        "require": "./dist/index.js"
      },
   ```
    - "types": "./dist/index.d.ts": This property specifies the location of the TypeScript type definitions for the package. 
    - "import": "./dist/index.mjs": This property specifies the path to the module that should be imported when using the ES module (import) syntax
    - "require": "./dist/index.js": This property specifies the path to the module that should be loaded when the package is required in a CommonJS context.
<br />
- peerDependencies: 
   - implication: specify the packages that are required by a npm package to work but are not automatically installed when this lib is installed, 
   - application scenarios:  building themes, plugins, or libraries as a supplement to another major lib. For example
      - make a plugin to a jQuery 
      - build a component system for react
      
-  **sideEffects** 

   > 
In the `package.json` file, the `"sideEffects"` field is used to indicate whether a package has side effects or not. It is a feature introduced in webpack 4 and is also supported by other build tools like Rollup.

The purpose of the `"sideEffects"` field is to help build tools optimize the bundle size by allowing them to perform tree shaking more effectively. Tree shaking is a technique used to eliminate dead code from the final bundle.

Here's how the `"sideEffects"` field works:

1. When `"sideEffects"` is set to `false`:
   - It indicates that the package does not have any side effects.
   - This means that the package's modules can be safely removed if they are not actually used in the code.
   - The build tool can perform more aggressive tree shaking, removing unused exports and reducing the bundle size.

2. When `"sideEffects"` is set to an array of file paths:
   - It specifies a list of files within the package that have side effects.
   - The build tool will not perform tree shaking on these specified files, ensuring that their side effects are preserved.
   - All other files in the package that are not listed in the `"sideEffects"` array are considered side-effect-free and can be safely removed if unused.

3. When `"sideEffects"` is not specified or set to `true`:
   - It indicates that the package has side effects or the side effects are unknown.
   - The build tool will not perform any tree shaking optimizations on the package.
   - All files in the package will be included in the final bundle, regardless of whether they are actually used or not.

The `"sideEffects"` field is particularly useful for libraries and packages that are intended to be used as dependencies in other projects. By properly setting the `"sideEffects"` field, library authors can provide hints to build tools, enabling more efficient tree shaking and resulting in smaller bundle sizes for the consuming projects.

It's important to note that the `"sideEffects"` field relies on the accuracy of the package's code and the correct identification of side effects. If a package incorrectly specifies `"sideEffects": false` when it actually has side effects, it can lead to incorrect tree shaking and potential runtime errors.

To summarize, the `"sideEffects"` field in `package.json` is used to indicate whether a package has side effects or not. It helps build tools optimize the bundle size by performing more effective tree shaking, removing unused code, and reducing the overall bundle size of the consuming project.
    
<br />

- version:  package version, and please version with caution, here is the [guideline](https://semver.org/)



<br />
<br />

### Package Versioning 

- Guideline of semantic versioning, here is the [detail](https://semver.org/) 
- cli tool on versioning: 
    - doc: https://www.npmjs.com/package/@changesets/cli
    - tutorial in 2 minutes: https://www.youtube.com/watch?v=eh89VE3Mk5g&ab_channel=MattPocock

<br />
<br />

## Publish package to [NPM](https://npmjs.com/)

as private package.

 

### Go over above concepts with an example

[Sonner](https://github.com/emilkowalski/sonner), a npm package where you can learn concepts above.

- source file

![image](https://github.com/reboottime/lib-template/assets/84222273/c6950fc4-688b-4072-8f0d-44b79819384e)

- installed as a package

![image](https://github.com/reboottime/lib-template/assets/84222273/83a2de24-dec1-40ac-bc98-c69bb92f8f0f)

**Comment by reboottime** [comment]
## Design Patterns you may benefit from 

1. Patterns: https://javascriptpatterns.vercel.app/patterns/design-patterns/observer-pattern
2. Sample: https://clerk.com/docs/references/javascript/overview

**Comment by reboottime** [comment]
##  sideEffects in package.json


In the `package.json` file, the `"sideEffects"` field is used to indicate whether a package has side effects or not. It is a feature introduced in webpack 4 and is also supported by other build tools like Rollup.

The purpose of the `"sideEffects"` field is to help build tools optimize the bundle size by allowing them to perform tree shaking more effectively. Tree shaking is a technique used to eliminate dead code from the final bundle.

Here's how the `"sideEffects"` field works:

1. When `"sideEffects"` is set to `false`:
   - It indicates that the package does not have any side effects.
   - This means that the package's modules can be safely removed if they are not actually used in the code.
   - The build tool can perform more aggressive tree shaking, removing unused exports and reducing the bundle size.

2. When `"sideEffects"` is set to an array of file paths:
   - It specifies a list of files within the package that have side effects.
   - The build tool will not perform tree shaking on these specified files, ensuring that their side effects are preserved.
   - All other files in the package that are not listed in the `"sideEffects"` array are considered side-effect-free and can be safely removed if unused.

3. When `"sideEffects"` is not specified or set to `true`:
   - It indicates that the package has side effects or the side effects are unknown.
   - The build tool will not perform any tree shaking optimizations on the package.
   - All files in the package will be included in the final bundle, regardless of whether they are actually used or not.

The `"sideEffects"` field is particularly useful for libraries and packages that are intended to be used as dependencies in other projects. By properly setting the `"sideEffects"` field, library authors can provide hints to build tools, enabling more efficient tree shaking and resulting in smaller bundle sizes for the consuming projects.

It's important to note that the `"sideEffects"` field relies on the accuracy of the package's code and the correct identification of side effects. If a package incorrectly specifies `"sideEffects": false` when it actually has side effects, it can lead to incorrect tree shaking and potential runtime errors.

To summarize, the `"sideEffects"` field in `package.json` is used to indicate whether a package has side effects or not. It helps build tools optimize the bundle size by performing more effective tree shaking, removing unused code, and reducing the overall bundle size of the consuming project.
